include/deal.II/numerics/vector_tools_hermite.h

```diff
@@ -0,0 +1,133 @@
+#include <deal.II/base/types.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/mapping.h>
+
+#include <map>
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace VectorTools
+{
+  /**
+   * @name Domain and boundary projection for Hermite elements
+   * @{
+   */
+
+  /**
+   * @p Enumeration representing the type of boundary condition to be enforced.
+   * Since Hermite finite elements can strongly impose continuity in various
+   * derivatives as well as value at element boundaries, they can also be used
+   * to impose boundary conditions such as Neumann conditions directly to reduce
+   * the size of the computational system.
+   */
+  enum HermiteBoundaryType
+  {
+    hermite_dirichlet,
+    hermite_neumann,
+    hermite_2nd_derivative,
+  };
+
+
+
+  /**
+   * Enforces boundary conditions by projecting onto the Hermite finite element
+   * space at the boundary.
+   *
+   * This is required even in the case of Dirichlet boundary conditions where
+   * similar functions already exist for other bases, because Hermite assigns
+   * degrees of freedom directly onto element boundaries that have a zero shape
+   * value across entire element faces. For most existing methods this can
+   * create a singular mass matrix unless special steps are taken to remove
+   * those degrees of freedom from the boundary system.
+   *
+   * Note that currently only Dirichlet, Neumann and second degree Neumann
+   * conditions are currently supported.
+   */
+  template <int dim, int spacedim = dim, typename Number = double>
+  void
+  project_hermite_boundary_values(
+    const Mapping<dim, spacedim> &   mapping_h,
+    const DoFHandler<dim, spacedim> &dof_handler,
+    const std::map<types::boundary_id, const Function<spacedim, Number> *>
+      &                                        boundary_functions,
+    const Quadrature<dim - 1> &                quadrature,
+    const HermiteBoundaryType                  projection_mode,
+    std::map<types::global_dof_index, Number> &boundary_values,
+    std::vector<unsigned int>                  component_mapping = {});
+
+
+
+  /**
+   * The following does exactly the same as the above, but uses
+   * hermite_dirichlet as the value
+   * for @p projection_mode.
+   */
+  template <int dim, int spacedim = dim, typename Number = double>
+  void
+  project_hermite_boundary_values(
+    const Mapping<dim, spacedim> &   mapping_h,
+    const DoFHandler<dim, spacedim> &dof_handler,
+    const std::map<types::boundary_id, const Function<spacedim, Number> *>
+      &                                        boundary_functions,
+    const Quadrature<dim - 1> &                quadrature,
+    std::map<types::global_dof_index, Number> &boundary_values,
+    std::vector<unsigned int>                  component_mapping = {});
+
+
+
+  /**
+   * Projection function for use with Hermite finite elements. Usually the existing @p project
+   * function will work with Hermite, however setting any of the flags for
```

Comment:
Question: have you added asserts there (so that one does not use the function if boundary terms are enabled)?